server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

```diff
@@ -553,13 +592,22 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 builder.field("path_unmatch", pathUnmatch);
             }
         }
-        // We have more than one types when (1) `match_mapping_type` is "*", and (2) match and/or path_match are defined but
-        // not `match_mapping_type`. In the latter the template implicitly accepts all types and we don't need to serialize
-        // the `match_mapping_type` values.
-        if (xContentFieldTypes.length > 1 && match.isEmpty() && pathMatch.isEmpty()) {
-            builder.field("match_mapping_type", "*");
-        } else if (xContentFieldTypes.length == 1) {
+        // If we can match all types (considering runtime support), then we can skip serializing "match_mapping_type".
+        if (xContentFieldTypes.length == 1) {
             builder.field("match_mapping_type", xContentFieldTypes[0]);
+        } else if (xContentFieldTypes.length != 0) {
+            final long numPossibleXContentFieldTypes;
+            if (runtimeMapping) {
+                numPossibleXContentFieldTypes = List.of(XContentFieldType.values())
+                    .stream()
+                    .filter(XContentFieldType::supportsRuntimeField)
+                    .count();
+            } else {
+                numPossibleXContentFieldTypes = XContentFieldType.values().length;
+            }
+            if (xContentFieldTypes.length < numPossibleXContentFieldTypes) {
+                builder.field("match_mapping_type", List.of(xContentFieldTypes).stream().map(XContentFieldType::toString).toList());
```

Comment:
Seems like if you set `unmatch_mapping_type` and then get the mapping, it'll be modified to a `match_mapping_type`. I think we'd generally want to return the mapping as it has been provided by the user, without modifying it.

Reply:
Agreed. I've fixed that -- I'm now passing in the user-specified `match_mapping_type` and `unmatch_mapping_type` lists, and serialising those. The final filtered list of content types is not used in the serialisation.

I have also removed the bit about skipping serialisation of `match_mapping_type` -- it will still be skipped for the implicit wildcard (because that is not touching the original lists). If a user explicitly specifies `match_mapping_type: "*"`, then that will now be serialised. Feels better/less magical to me.